### PR TITLE
make scope tree listeners more event driven

### PR DESCRIPTION
**Pull request by kevinushey**
This further refines the tokenization improvements by making the background tokenization fire an update event on completion, which clients can listen to as a signal for updating (ie, the status bar and the document outline widget).

@jmcphers, if you have time later can we review together?

### CODE REVIEW

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java [hunk 1]
```diff
@@ -2920,41 +2930,75 @@ private void fireLineWidgetsChanged()
...
+      
+      private final AceEditor editor_;
       private final Timer timer_;
+      
       private int row_ = 0;
+      private boolean mutated_ = true;
       
-      private static final int DELAY_MS = 5;
-      private static final int ROWS_TOKENIZED_PER_ITERATION = 50;
+      private static final int DELAY_MS = 0;
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java]: Note that this change causes much more eager background tokenization, but the background tokenizer stops as soon as the user starts moving around in the editor (cursor movements etc.).

I've been tweaking this as 100 just seemed too slow in larger documents and didn't confer the benefits we'd hope to get from background tokenization. The downside is that if the user starts editing while the background tokenizer is running in a large document, there will be a brief ~500ms pause until that particular iteration finishes.

Comment (jjallaire) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java]: I don't love the idea of 500ms delay under any circumstances.

**Comment by jjallaire** [comment]
I'm assuming that this work is for vNext and the previous work is okay going into the patch release? (i.e. these PRs are independent).

**Comment by kevinushey** [comment]
Yep, that's right (this is just refinement on top of the previous PR)

**Comment by jjallaire** [comment]
So we can take the previous one in the patch release and not take this one?
(since it's more risky/intrusive)

**Comment by kevinushey** [comment]
Yes, that's correct -- let's take the previous work for the patch release. (sorry, I should have made that more clear!)

**Comment by jjallaire** [comment]
Okay, sounds good!

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java (hunk 1 above)
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java]: I'll play around more to see if I can find a better solution.

**Comment by kevinushey** [comment]
Turns out there was some sorely inefficient tokenization being incurred by the token iterator (an error in updating the last tokenized row was causing eager re-tokenization of the whole document... again!).

With that fixed, we can easily do eager background tokenization without impacting the editor experience, even in large documents.

**Comment by jjallaire** [comment]
Awesome!

**Comment by kevinushey** [comment]
I think this one is good to go (it should greatly improve the tokenization performance in large documents)

**Comment by jjallaire** [comment]
@jmcphers Could you review this also and merge if all looks well?

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java [hunk 2]
```diff
@@ -2920,41 +2930,55 @@ private void fireLineWidgetsChanged()
...
-         editor.addCursorChangedHandler(new CursorChangedHandler()
+         editor_.addDocumentChangedHandler(new DocumentChangedEvent.Handler()
          {
             @Override
-            public void onCursorChanged(CursorChangedEvent event)
+            public void onDocumentChanged(DocumentChangedEvent event)
             {
-               int row = event.getPosition().getRow();
-               row_ = row;
-               timer_.schedule(editor.getSuggestedScopeUpdateDelay() / 2);
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java]: Can you clean up `getSuggestedScopeUpdateDelay` if we don't need it any more?

**Comment by jmcphers** [comment]
Looks good @kevinushey.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java (hunk 2 above)
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java]: Will do!

**Pull request merged**